pangeo/pbs.py

```diff
@@ -0,0 +1,163 @@
+from contextlib import contextmanager
+import logging
+import os
+import socket
+import subprocess
+import sys
+
+from distributed import LocalCluster
+from distributed.utils import tmpfile, get_ip_interface
+
+template = """
+#!/bin/bash
+#PBS -N %(name)s
+#PBS -q %(queue)s
+#PBS -A %(project)s
+#PBS -l select=1:ncpus=%(threads_per_worker)d:mpiprocs=1:ompthreads=1
+#PBS -l walltime=%(walltime)s
+#PBS -j oe
+#PBS -m abe
+
+%(base_path)s/dask-worker %(scheduler)s --nthreads %(threads_per_worker)d --memory-limit %(memory)s --name $PBS_JOBNAME-$PBS_JOBID-$PBS_TASKNUM %(extra)s
+"""
+
+
+logger = logging.getLogger(__name__)
+
+
+dirname = os.path.dirname(sys.executable)
+
+
+class PBSCluster(object):
+    """ Launch Dask on a PBS cluster
+
+    Examples
+    --------
+    >>> from pangeo import PBSCluster
+    >>> cluster = PBSCluster(project='...')
+    >>> cluster.start_workers(10)  # this may take a few seconds to launch
+
+    >>> from dask.distributed import Client
+    >>> client = Client(cluster)
+
+    This also works with adaptive clusters.  This automatically launches and
+    kill workers based on load.
+
+    >>> from distributed.deploy import Adaptive
+    >>> adapt = Adaptive(cluster.cluster.scheduler, cluster)
+    """
+    def __init__(self,
+                 name='dask',
+                 q='regular',
+                 project='UCLB0022',  # is there an environment variable I can use?
```

Comment:
Is there a way to obtain a default project for a particular user?  @davidedelvento

Reply:
PBS does not have a concept of default project, however NCAR added it for users who wish to set it up for their own reasons, by using the `$PBS_ACCOUNT` environmental variable.

Some users set it, and it'd be great if you'd use it for them, but many users do not set it. So you need to cover for when it's missing. In this case, you can pick the first project they have or one at random. See how we (== the group I work in, aka CSG) implemented this logic in `/glade/u/apps/ch/opt/usr/bin/qcmd` and `/glade/u/apps/ch/opt/usr/bin/qinteractive` (they are bash scripts, IIRC). Both of them (IIRC) rely on `/ncar/opt/sam/sam_validate.py` which is out of the control of CSG (as you can see it's owned by root, which we don't have). Also, you can see this thing is not truly made for importing, but I could see if I find who maintains it and put you in touch with them if you prefer to do use a stable API vs parsing text as bash does.